libs/full/async_mpi/include/hpx/async_mpi/transform_mpi.hpp

```diff
@@ -0,0 +1,143 @@
+                            "failed with MPI error code: ", mpi_status)));
+            }
+        }
+
+        template <typename R, typename... Ts>
+        void set_value_request_callback(MPI_Request request, R&& r, Ts&&... ts)
+        {
+            detail::add_request_callback(
+                [r = std::forward<R>(r),
+                    keep_alive = hpx::make_tuple(std::forward<Ts>(ts)...)](
+                        int status) mutable
+                    {
+                        set_value_request_callback_helper(std::move(r), status);
+                    },
+                request);
+        }
+
+        template <typename R, typename F>
+        struct transform_mpi_receiver
+        {
+            std::decay_t<R> r;
+            std::decay_t<F> f;
+
+            template <typename R_, typename F_>
+            transform_mpi_receiver(R_&& r, F_&& f)
+              : r(std::forward<R_>(r)), f(std::forward<F_>(f))
+            {
+            }
+
+            template <typename E>
+            void set_error(E&& e) noexcept
+            {
+                hpx::execution::experimental::set_error(
+                    std::move(r), std::forward<E>(e));
+            }
+
+            void set_done() noexcept
+            {
+                hpx::execution::experimental::set_done(std::move(r));
+            };
+
+            template <typename... Ts,
+                     typename = std::enable_if_t<hpx::is_invocable_v<F, Ts..., MPI_Request*>>>
+            void set_value(Ts&&... ts) && noexcept
+            {
+                hpx::detail::try_catch_exception_ptr(
+                    [&]() {
+                        MPI_Request request;
+                        HPX_INVOKE(f, std::forward<Ts>(ts)..., &request);
+                        // Return value discarded as error assessed when
+                        // doing HPX_Testany on the request
+                        set_value_request_callback(request,
+                                std::move(r), std::forward<Ts>(ts)...);
+                    },
+                    [&](std::exception_ptr ep) {
+                        hpx::execution::experimental::set_error(
+                            std::move(r), std::move(ep));
+                    });
+            }
+        };
+
+        template <typename Sender, typename F>
+        struct transform_mpi_sender
+        {
+            std::decay_t<Sender> s;
+            std::decay_t<F> f;
+
+            // TODO: Handle all types of return values and not only int as we
+            // can pass lambdas
+            template <template <typename...> class Tuple,
+                template <typename...> class Variant>
+            using value_types = Variant<Tuple<int>>;
+
+            template <template <typename...> class Variant>
+            using error_types =
+                typename hpx::execution::experimental::sender_traits<
+                    Sender>::template error_types<Variant>;
```

Comment:
This should add `exception_ptr` to the list. See `transform_stream` for an example of how to do that.